pkg/phases/kernel.go

```diff
@@ -0,0 +1,127 @@
+package phases
+
+import (
+	"fmt"
+        "strings"
+
+	//log "github.com/sirupsen/logrus"
+
+	"github.com/flanksource/konfigadm/pkg/types"
+	"github.com/flanksource/konfigadm/pkg/utils"
+)
+
+var Kernel types.AllPhases = kernel{}
+
+type kernel struct{}
+
+func (p kernel) ApplyPhase(sys *types.Config, ctx *types.SystemContext) ([]types.Command, types.Filesystem, error) {
+	commands := types.Commands{}
+	files := types.Filesystem{}
+	var os OS
+	var err error
+        if sys.Kernel == "" {
+		return []types.Command{}, files, nil
+	}
+	if os, err = GetOSForTag(ctx.Flags...); err != nil {
+		return nil, nil, err
+        }
+        kernelpackage, kernelheaderpackage := os.GetKernelPackageNames(sys.Kernel)
+        commands.Append(os.GetPackageManager().Install(kernelpackage, kernelheaderpackage))
+        // Possible Enhancement, implement this logic in an os.GetGrubManager()
+        return commands.Merge(), files, nil
+}
+
+func (p kernel) Verify(cfg *types.Config, results *types.VerifyResults, flags ...types.Flag) bool {
+	verify := true
+	var os OS
+        var err error
+	if os, err = GetOSForTag(flags...); err != nil {
+		results.Fail("Unable to find OS for tags %s", flags)
+		return false
+	}
+        if cfg.Kernel == "" {
+		return verify
+	}
+        kernelpackage, kernelheaderpackage := os.GetKernelPackageNames(cfg.Kernel)
+        installed := os.GetPackageManager().GetInstalledVersion(kernelpackage)
+        if strings.Contains(installed, cfg.Kernel) {
+		results.Pass("%s is installed", kernelpackage)
+        } else {
+		results.Fail("%s is not installed", kernelpackage)
+                verify = verify && false
+        }
+        installed = os.GetPackageManager().GetInstalledVersion(kernelheaderpackage)
+        if strings.Contains(installed, cfg.Kernel) {
+		results.Pass("%s is installed", kernelheaderpackage)
+        } else {
+		results.Fail("%s is not installed", kernelheaderpackage)
+                verify = verify && false
+        }
+        if os.VerifyDefaultKernel(cfg.Kernel) {
+		results.Pass("%s is the default kernel", cfg.Kernel)
+        } else {
+		results.Fail("%s is not the default kernel", cfg.Kernel)
+                verify = verify && false
+        }
+
+	return verify
+}
+func (p kernel) ProcessFlags(sys *types.Config, flags ...types.Flag) {
+	//All the stuff is done in commands minification?
+}
+
+//This interface and the two corresponding structs are used to provide OS interface easy interface to the common naming conventions
+type KernelNameManager interface {
+        GetKernelPackageNames(version string) (string, string)
```

Comment:
This method can probably just move into the OS interface/structs

Reply:
I decided to absract out the idea of being type 'linux' vs type 'kernel', mostly to avoid code repetition.  Maybe overkill given how unlikely it is for the package names to ever change in the future?  The OS does also present a GetKernelPackageName which allows each OS to call the relevant KernelNameManager.  The other option would maybe be for the kernel name to be an interface in yum/apt since they correlate in all of the supported OSs.  In future I'd also like the yum based kernels to auto fill in %{?dist}  and uname -p instead of needing the user to ensure that's in the input string.

Reply:
> other option would maybe be for the kernel name to be an interface in yum/apt since they correlate in all of the supported OSs

👍 

> In future I'd also like the yum based kernels to auto fill in %{?dist} and uname -p 

In theory, you can do something like this: https://github.com/flanksource/konfigadm/blob/master/pkg/apps/cri.go#L155